# EXPERIMENT NO 1: Data preparation using NumPy and Pandas

Dataset Link : https://www.kaggle.com/datasets/rajeshrampure/black-friday-sale

# 1. Load Data in Pandas

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('train.csv')

# 2. Description of the dataset.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Product_Category_3          166821 non-null  float64
 11  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


In [ ]:
df.describe()

User_ID     Occupation  Marital_Status  Product_Category_1  \
count  5.500680e+05  550068.000000   550068.000000       550068.000000   
mean   1.003029e+06       8.076707        0.409653            5.404270   
std    1.727592e+03       6.522660        0.491770            3.936211   
min    1.000001e+06       0.000000        0.000000            1.000000   
25%    1.001516e+06       2.000000        0.000000            1.000000   
50%    1.003077e+06       7.000000        0.000000            5.000000   
75%    1.004478e+06      14.000000        1.000000            8.000000   
max    1.006040e+06      20.000000        1.000000           20.000000   

       Product_Category_2  Product_Category_3       Purchase  
count       376430.000000       166821.000000  550068.000000  
mean             9.842329           12.668243    9263.968713  
std              5.086590            4.125338    5023.065394  
min              2.000000            3.000000      12.000000  
25%              5.000000            9.000000    5823.000000  
50%              9.000000           14.000000    8047.000000  
75%             15.000000           16.000000   12054.000000  
max             18.000000           18.000000   23961.000000

In [ ]:
df.shape

(550068, 12)

In [ ]:
df.head(20)

User_ID Product_ID Gender    Age  Occupation City_Category  \
0   1000001  P00069042      F   0-17          10             A   
1   1000001  P00248942      F   0-17          10             A   
2   1000001  P00087842      F   0-17          10             A   
3   1000001  P00085442      F   0-17          10             A   
4   1000002  P00285442      M    55+          16             C   
5   1000003  P00193542      M  26-35          15             A   
6   1000004  P00184942      M  46-50           7             B   
7   1000004  P00346142      M  46-50           7             B   
8   1000004   P0097242      M  46-50           7             B   
9   1000005  P00274942      M  26-35          20             A   
10  1000005  P00251242      M  26-35          20             A   
11  1000005  P00014542      M  26-35          20             A   
12  1000005  P00031342      M  26-35          20             A   
13  1000005  P00145042      M  26-35          20             A   
14  1000006  P00231342      F  51-55           9             A   
15  1000006  P00190242      F  51-55           9             A   
16  1000006   P0096642      F  51-55           9             A   
17  1000006  P00058442      F  51-55           9             A   
18  1000007  P00036842      M  36-45           1             B   
19  1000008  P00249542      M  26-35          12             C   

   Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                           2               0                   3   
1                           2               0                   1   
2                           2               0                  12   
3                           2               0                  12   
4                          4+               0                   8   
5                           3               0                   1   
6                           2               1                   1   
7                           2               1                   1   
8                           2               1                   1   
9                           1               1                   8   
10                          1               1                   5   
11                          1               1                   8   
12                          1               1                   8   
13                          1               1                   1   
14                          1               0                   5   
15                          1               0                   4   
16                          1               0                   2   
17                          1               0                   5   
18                          1               1                   1   
19                         4+               1                   1   

    Product_Category_2  Product_Category_3  Purchase  
0                  NaN                 NaN      8370  
1                  6.0                14.0     15200  
2                  NaN                 NaN      1422  
3                 14.0                 NaN      1057  
4                  NaN                 NaN      7969  
5                  2.0                 NaN     15227  
6                  8.0                17.0     19215  
7                 15.0                 NaN     15854  
8                 16.0                 NaN     15686  
9                  NaN                 NaN      7871  
10                11.0                 NaN      5254  
11                 NaN                 NaN      3957  
12                 NaN                 NaN      6073  
13                 2.0                 5.0     15665  
14                 8.0                14.0      5378  
15                 5.0                 NaN      2079  
16                 3.0                 4.0     13055  
17                14.0                 NaN      8851  
18                14.0                16.0     11788  
19                 5.0                15.0     19614

# 3. Drop columns that aren’t useful.

 We will remove the "User_ID" column since it doesn't contribute to predicting the target variable (Purchase).

In [ ]:
df.drop('User_ID', axis=1, inplace=True)

In [ ]:
df.head(1)

Product_ID Gender   Age  Occupation City_Category  \
0  P00069042      F  0-17          10             A   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 11 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Product_ID                  550068 non-null  object 
 1   Gender                      550068 non-null  object 
 2   Age                         550068 non-null  object 
 3   Occupation                  550068 non-null  int64  
 4   City_Category               550068 non-null  object 
 5   Stay_In_Current_City_Years  550068 non-null  object 
 6   Marital_Status              550068 non-null  int64  
 7   Product_Category_1          550068 non-null  int64  
 8   Product_Category_2          376430 non-null  float64
 9   Product_Category_3          166821 non-null  float64
 10  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(4), object(5)
memory usage: 46.2+ MB


# 4. Drop rows with maximum missing values.

In [ ]:
df.shape

(550068, 11)

In [ ]:
df.isnull().sum()

Product_ID                         0
Gender                             0
Age                                0
Occupation                         0
City_Category                      0
Stay_In_Current_City_Years         0
Marital_Status                     0
Product_Category_1                 0
Product_Category_2            173638
Product_Category_3            383247
Purchase                           0
dtype: int64

Checking Unique values for Product_category

In [ ]:
df['Product_Category_1'].unique()

array([ 3,  1, 12,  8,  5,  4,  2,  6, 14, 11, 13, 15,  7, 16, 18, 10, 17,
        9, 20, 19])

In [ ]:
df['Product_Category_2'].unique()

array([nan,  6., 14.,  2.,  8., 15., 16., 11.,  5.,  3.,  4., 12.,  9.,
       10., 17., 13.,  7., 18.])

In [ ]:
df['Product_Category_3'].unique()

array([nan, 14., 17.,  5.,  4., 16., 15.,  8.,  9., 13.,  6., 12.,  3.,
       18., 11., 10.])

Lets skip the above process as some products can be of only one category.This ensures no information is lost and allows products belonging to single and multiple categories to be represented properly(can be converted into dummy variables)

# 5. Take care of missing data.

Lets assume 0 to be representing 'no-category'.Hence, fill missing values in Product_Category_2 and Product_Category_3 with 0 .

In [ ]:
# Fill missing values in Product_Category_2 and Product_Category_3 with 0
df['Product_Category_2'] = df['Product_Category_2'].fillna(0)
df['Product_Category_3'] = df['Product_Category_3'].fillna(0)

# Convert Product_Category_2 and Product_Category_3 to integer type
df['Product_Category_2'] = df['Product_Category_2'].astype(int)
df['Product_Category_3'] = df['Product_Category_3'].astype(int)

# Verify the changes
df[['Product_Category_2', 'Product_Category_3']].head()


Product_Category_2  Product_Category_3
0                   0                   0
1                   6                  14
2                   0                   0
3                  14                   0
4                   0                   0

In [ ]:
df.isnull().sum()

Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            0
Product_Category_3            0
Purchase                      0
dtype: int64

In [ ]:
df.shape

(550068, 11)

# 6. Create dummy variables.       

In [ ]:
df.head()

Product_ID Gender   Age  Occupation City_Category  \
0  P00069042      F  0-17          10             A   
1  P00248942      F  0-17          10             A   
2  P00087842      F  0-17          10             A   
3  P00085442      F  0-17          10             A   
4  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                   0                   0      8370  
1                   6                  14     15200  
2                   0                   0      1422  
3                  14                   0      1057  
4                   0                   0      7969

In [ ]:
# Strip any leading/trailing spaces in column names
df.columns = df.columns.str.strip()

# Now, run the value_counts for Product_Category_1
product_counts = df['Product_Category_1'].value_counts()
print(product_counts)


Product_Category_1
5     150933
1     140378
8     113925
11     24287
2      23864
6      20466
3      20213
4      11753
16      9828
15      6290
13      5549
10      5125
12      3947
7       3721
18      3125
20      2550
19      1603
14      1523
17       578
9        410
Name: count, dtype: int64


In [ ]:
age_counts = df['Age'].value_counts()
print(age_counts)

Age
26-35    219587
36-45    110013
18-25     99660
46-50     45701
51-55     38501
55+       21504
0-17      15102
Name: count, dtype: int64


In [ ]:
age_counts = df['Gender'].value_counts()
print(age_counts)

Gender
M    414259
F    135809
Name: count, dtype: int64


In [ ]:
age_counts = df['Stay_In_Current_City_Years'].value_counts()
print(age_counts)

Stay_In_Current_City_Years
1     193821
2     101838
3      95285
4+     84726
0      74398
Name: count, dtype: int64


In [ ]:
age_counts = df['Occupation'].value_counts()
print(age_counts)

Occupation
4     72308
0     69638
7     59133
1     47426
17    40043
20    33562
12    31179
14    27309
2     26588
16    25371
6     20355
3     17650
10    12930
5     12177
15    12165
11    11586
19     8461
13     7728
18     6622
9      6291
8      1546
Name: count, dtype: int64


In [ ]:
# Create dummy variables for categorical columns
df_dummies = pd.get_dummies(df, columns=['Gender', 'Age', 'City_Category', 'Stay_In_Current_City_Years'], drop_first=True)

# Convert only boolean columns to integers
for col in df_dummies.select_dtypes(include=['bool']).columns:
    df_dummies[col] = df_dummies[col].astype(int)

# Check the updated dataframe
df_dummies.head()


Product_ID  Occupation  Marital_Status  Product_Category_1  \
0  P00069042          10               0                   3   
1  P00248942          10               0                   1   
2  P00087842          10               0                  12   
3  P00085442          10               0                  12   
4  P00285442          16               0                   8   

   Product_Category_2  Product_Category_3  Purchase  Gender_M  Age_18-25  \
0                   0                   0      8370         0          0   
1                   6                  14     15200         0          0   
2                   0                   0      1422         0          0   
3                  14                   0      1057         0          0   
4                   0                   0      7969         1          0   

   Age_26-35  Age_36-45  Age_46-50  Age_51-55  Age_55+  City_Category_B  \
0          0          0          0          0        0                0   
1          0          0          0          0        0                0   
2          0          0          0          0        0                0   
3          0          0          0          0        0                0   
4          0          0          0          0        1                0   

   City_Category_C  Stay_In_Current_City_Years_1  \
0                0                             0   
1                0                             0   
2                0                             0   
3                0                             0   
4                1                             0   

   Stay_In_Current_City_Years_2  Stay_In_Current_City_Years_3  \
0                             1                             0   
1                             1                             0   
2                             1                             0   
3                             1                             0   
4                             0                             0   

   Stay_In_Current_City_Years_4+  
0                              0  
1                              0  
2                              0  
3                              0  
4                              1

 Product_Category might be unordered so coverting into dummy variable.(Whereas occupation is ordered)

In [ ]:
df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 20 columns):
 #   Column                         Non-Null Count   Dtype 
---  ------                         --------------   ----- 
 0   Product_ID                     550068 non-null  object
 1   Occupation                     550068 non-null  int64 
 2   Marital_Status                 550068 non-null  int64 
 3   Product_Category_1             550068 non-null  int64 
 4   Product_Category_2             550068 non-null  int64 
 5   Product_Category_3             550068 non-null  int64 
 6   Purchase                       550068 non-null  int64 
 7   Gender_M                       550068 non-null  int64 
 8   Age_18-25                      550068 non-null  int64 
 9   Age_26-35                      550068 non-null  int64 
 10  Age_36-45                      550068 non-null  int64 
 11  Age_46-50                      550068 non-null  int64 
 12  Age_51-55                      550068 non-nu

In [ ]:
import pandas as pd

# Assuming df_dummies already exists
# Step 1: Define the range of categories
categories = range(21)  # Categories from 0 to 20

# Step 2: Initialize an empty DataFrame with required columns
df_dummies1 = pd.DataFrame(0, index=df_dummies.index, columns=[f'category_{i}' for i in categories])

# Step 3: Iterate over Product_Category columns and update df_dummies1
for col in ['Product_Category_1', 'Product_Category_2', 'Product_Category_3']:
    for category in categories:
        df_dummies1[f'category_{category}'] |= (df_dummies[col] == category).astype(int)


In [ ]:
df_dummies1.head()

category_0  category_1  category_2  category_3  category_4  category_5  \
0           1           0           0           1           0           0   
1           0           1           0           0           0           0   
2           1           0           0           0           0           0   
3           1           0           0           0           0           0   
4           1           0           0           0           0           0   

   category_6  category_7  category_8  category_9  ...  category_11  \
0           0           0           0           0  ...            0   
1           1           0           0           0  ...            0   
2           0           0           0           0  ...            0   
3           0           0           0           0  ...            0   
4           0           0           1           0  ...            0   

   category_12  category_13  category_14  category_15  category_16  \
0            0            0            0            0            0   
1            0            0            1            0            0   
2            1            0            0            0            0   
3            1            0            1            0            0   
4            0            0            0            0            0   

   category_17  category_18  category_19  category_20  
0            0            0            0            0  
1            0            0            0            0  
2            0            0            0            0  
3            0            0            0            0  
4            0            0            0            0  

[5 rows x 21 columns]

In [ ]:
df_dummies.head()

Product_ID  Occupation  Marital_Status  Product_Category_1  \
0  P00069042          10               0                   3   
1  P00248942          10               0                   1   
2  P00087842          10               0                  12   
3  P00085442          10               0                  12   
4  P00285442          16               0                   8   

   Product_Category_2  Product_Category_3  Purchase  Gender_M  Age_18-25  \
0                   0                   0      8370         0          0   
1                   6                  14     15200         0          0   
2                   0                   0      1422         0          0   
3                  14                   0      1057         0          0   
4                   0                   0      7969         1          0   

   Age_26-35  Age_36-45  Age_46-50  Age_51-55  Age_55+  City_Category_B  \
0          0          0          0          0        0                0   
1          0          0          0          0        0                0   
2          0          0          0          0        0                0   
3          0          0          0          0        0                0   
4          0          0          0          0        1                0   

   City_Category_C  Stay_In_Current_City_Years_1  \
0                0                             0   
1                0                             0   
2                0                             0   
3                0                             0   
4                1                             0   

   Stay_In_Current_City_Years_2  Stay_In_Current_City_Years_3  \
0                             1                             0   
1                             1                             0   
2                             1                             0   
3                             1                             0   
4                             0                             0   

   Stay_In_Current_City_Years_4+  
0                              0  
1                              0  
2                              0  
3                              0  
4                              1

In [ ]:
# View all columns of df_dummies
pd.set_option('display.max_columns', None)
df_dummies.head()


Product_ID  Occupation  Marital_Status  Product_Category_1  \
0  P00069042          10               0                   3   
1  P00248942          10               0                   1   
2  P00087842          10               0                  12   
3  P00085442          10               0                  12   
4  P00285442          16               0                   8   

   Product_Category_2  Product_Category_3  Purchase  Gender_M  Age_18-25  \
0                   0                   0      8370         0          0   
1                   6                  14     15200         0          0   
2                   0                   0      1422         0          0   
3                  14                   0      1057         0          0   
4                   0                   0      7969         1          0   

   Age_26-35  Age_36-45  Age_46-50  Age_51-55  Age_55+  City_Category_B  \
0          0          0          0          0        0                0   
1          0          0          0          0        0                0   
2          0          0          0          0        0                0   
3          0          0          0          0        0                0   
4          0          0          0          0        1                0   

   City_Category_C  Stay_In_Current_City_Years_1  \
0                0                             0   
1                0                             0   
2                0                             0   
3                0                             0   
4                1                             0   

   Stay_In_Current_City_Years_2  Stay_In_Current_City_Years_3  \
0                             1                             0   
1                             1                             0   
2                             1                             0   
3                             1                             0   
4                             0                             0   

   Stay_In_Current_City_Years_4+  
0                              0  
1                              0  
2                              0  
3                              0  
4                              1

In [ ]:
df_dummies.shape

(550068, 20)

In [ ]:
# Drop the original Product_Category columns from df_dummies
df_dummies = df_dummies.drop(columns=['Product_Category_1', 'Product_Category_2', 'Product_Category_3'])

# Drop the 'category_0' column from df_dummies1
df_dummies1 = df_dummies1.drop(columns=['category_0'])

# Concatenate the two dataframes along the columns
df_1 = pd.concat([df_dummies, df_dummies1], axis=1)

# Display the resulting dataframe
df_1.head()


Product_ID  Occupation  Marital_Status  Purchase  Gender_M  Age_18-25  \
0  P00069042          10               0      8370         0          0   
1  P00248942          10               0     15200         0          0   
2  P00087842          10               0      1422         0          0   
3  P00085442          10               0      1057         0          0   
4  P00285442          16               0      7969         1          0   

   Age_26-35  Age_36-45  Age_46-50  Age_51-55  Age_55+  City_Category_B  \
0          0          0          0          0        0                0   
1          0          0          0          0        0                0   
2          0          0          0          0        0                0   
3          0          0          0          0        0                0   
4          0          0          0          0        1                0   

   City_Category_C  Stay_In_Current_City_Years_1  \
0                0                             0   
1                0                             0   
2                0                             0   
3                0                             0   
4                1                             0   

   Stay_In_Current_City_Years_2  Stay_In_Current_City_Years_3  \
0                             1                             0   
1                             1                             0   
2                             1                             0   
3                             1                             0   
4                             0                             0   

   Stay_In_Current_City_Years_4+  category_1  category_2  category_3  \
0                              0           0           0           1   
1                              0           1           0           0   
2                              0           0           0           0   
3                              0           0           0           0   
4                              1           0           0           0   

   category_4  category_5  category_6  category_7  category_8  category_9  \
0           0           0           0           0           0           0   
1           0           0           1           0           0           0   
2           0           0           0           0           0           0   
3           0           0           0           0           0           0   
4           0           0           0           0           1           0   

   category_10  category_11  category_12  category_13  category_14  \
0            0            0            0            0            0   
1            0            0            0            0            1   
2            0            0            1            0            0   
3            0            0            1            0            1   
4            0            0            0            0            0   

   category_15  category_16  category_17  category_18  category_19  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   category_20  
0            0  
1            0  
2            0  
3            0  
4            0

In [ ]:
df_1.shape

(550068, 37)

new

In [ ]:
df_1['category_12'].value_counts()

category_12
0    531347
1     18721
Name: count, dtype: int64

In [ ]:
df_dummies.shape

(550068, 17)

In [ ]:
df.head()

Product_ID Gender   Age  Occupation City_Category  \
0  P00069042      F  0-17          10             A   
1  P00248942      F  0-17          10             A   
2  P00087842      F  0-17          10             A   
3  P00085442      F  0-17          10             A   
4  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                   0                   0      8370  
1                   6                  14     15200  
2                   0                   0      1422  
3                  14                   0      1057  
4                   0                   0      7969

In [ ]:
df_dummies.head()

Product_ID  Occupation  Marital_Status  Purchase  Gender_M  Age_18-25  \
0  P00069042          10               0      8370         0          0   
1  P00248942          10               0     15200         0          0   
2  P00087842          10               0      1422         0          0   
3  P00085442          10               0      1057         0          0   
4  P00285442          16               0      7969         1          0   

   Age_26-35  Age_36-45  Age_46-50  Age_51-55  Age_55+  City_Category_B  \
0          0          0          0          0        0                0   
1          0          0          0          0        0                0   
2          0          0          0          0        0                0   
3          0          0          0          0        0                0   
4          0          0          0          0        1                0   

   City_Category_C  Stay_In_Current_City_Years_1  \
0                0                             0   
1                0                             0   
2                0                             0   
3                0                             0   
4                1                             0   

   Stay_In_Current_City_Years_2  Stay_In_Current_City_Years_3  \
0                             1                             0   
1                             1                             0   
2                             1                             0   
3                             1                             0   
4                             0                             0   

   Stay_In_Current_City_Years_4+  
0                              0  
1                              0  
2                              0  
3                              0  
4                              1

# 7. Standardization and Normalization of columns

In [ ]:
df_1.describe()

Occupation  Marital_Status       Purchase       Gender_M  \
count  550068.000000   550068.000000  550068.000000  550068.000000   
mean        8.076707        0.409653    9263.968713       0.753105   
std         6.522660        0.491770    5023.065394       0.431205   
min         0.000000        0.000000      12.000000       0.000000   
25%         2.000000        0.000000    5823.000000       1.000000   
50%         7.000000        0.000000    8047.000000       1.000000   
75%        14.000000        1.000000   12054.000000       1.000000   
max        20.000000        1.000000   23961.000000       1.000000   

           Age_18-25      Age_26-35      Age_36-45      Age_46-50  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.181178       0.399200       0.199999       0.083082   
std         0.385166       0.489734       0.400000       0.276007   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

           Age_51-55        Age_55+  City_Category_B  City_Category_C  \
count  550068.000000  550068.000000    550068.000000    550068.000000   
mean        0.069993       0.039093         0.420263         0.311189   
std         0.255136       0.193817         0.493601         0.462980   
min         0.000000       0.000000         0.000000         0.000000   
25%         0.000000       0.000000         0.000000         0.000000   
50%         0.000000       0.000000         0.000000         0.000000   
75%         0.000000       0.000000         1.000000         1.000000   
max         1.000000       1.000000         1.000000         1.000000   

       Stay_In_Current_City_Years_1  Stay_In_Current_City_Years_2  \
count                 550068.000000                 550068.000000   
mean                       0.352358                      0.185137   
std                        0.477705                      0.388409   
min                        0.000000                      0.000000   
25%                        0.000000                      0.000000   
50%                        0.000000                      0.000000   
75%                        1.000000                      0.000000   
max                        1.000000                      1.000000   

       Stay_In_Current_City_Years_3  Stay_In_Current_City_Years_4+  \
count                 550068.000000                  550068.000000   
mean                       0.173224                       0.154028   
std                        0.378441                       0.360976   
min                        0.000000                       0.000000   
25%                        0.000000                       0.000000   
50%                        0.000000                       0.000000   
75%                        0.000000                       0.000000   
max                        1.000000                       1.000000   

          category_1     category_2     category_3     category_4  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.255201       0.132858       0.043104       0.071455   
std         0.435975       0.339422       0.203091       0.257583   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

          category_5     category_6     category_7     category_8  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.352367       0.076031       0.007903       0.346457   
std         0.477

**Columns to Standardize**

Purchase: This is a continuous variable with a large range (min: 12, max: 23961), so it should be standardized to have a mean of 0 and a standard deviation of 1.

**Columns to Normalize**
Categorical Columns (like Occupation, gender, Marital_Status,Stay_In_Current_City_Years_* , Product_Category_* columns): These are already in binary form (0 or 1) after creating the dummy variables, so they do not need normalization.

In [ ]:
df_1_noScaling = df_1.copy()

In [ ]:
from sklearn.preprocessing import StandardScaler

# Standardize 'Purchase'
scaler = StandardScaler()
df_1['Purchase'] = scaler.fit_transform(df_1[['Purchase']])
df_1.head()

Product_ID  Occupation  Marital_Status  Purchase  Gender_M  Age_18-25  \
0  P00069042          10               0 -0.177973         0          0   
1  P00248942          10               0  1.181756         0          0   
2  P00087842          10               0 -1.561193         0          0   
3  P00085442          10               0 -1.633858         0          0   
4  P00285442          16               0 -0.257805         1          0   

   Age_26-35  Age_36-45  Age_46-50  Age_51-55  Age_55+  City_Category_B  \
0          0          0          0          0        0                0   
1          0          0          0          0        0                0   
2          0          0          0          0        0                0   
3          0          0          0          0        0                0   
4          0          0          0          0        1                0   

   City_Category_C  Stay_In_Current_City_Years_1  \
0                0                             0   
1                0                             0   
2                0                             0   
3                0                             0   
4                1                             0   

   Stay_In_Current_City_Years_2  Stay_In_Current_City_Years_3  \
0                             1                             0   
1                             1                             0   
2                             1                             0   
3                             1                             0   
4                             0                             0   

   Stay_In_Current_City_Years_4+  category_1  category_2  category_3  \
0                              0           0           0           1   
1                              0           1           0           0   
2                              0           0           0           0   
3                              0           0           0           0   
4                              1           0           0           0   

   category_4  category_5  category_6  category_7  category_8  category_9  \
0           0           0           0           0           0           0   
1           0           0           1           0           0           0   
2           0           0           0           0           0           0   
3           0           0           0           0           0           0   
4           0           0           0           0           1           0   

   category_10  category_11  category_12  category_13  category_14  \
0            0            0            0            0            0   
1            0            0            0            0            1   
2            0            0            1            0            0   
3            0            0            1            0            1   
4            0            0            0            0            0   

   category_15  category_16  category_17  category_18  category_19  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   category_20  
0            0  
1            0  
2            0  
3            0  
4            0

In [ ]:
df_1.describe()


Occupation  Marital_Status      Purchase       Gender_M  \
count  550068.000000   550068.000000  5.500680e+05  550068.000000   
mean        8.076707        0.409653 -1.297937e-16       0.753105   
std         6.522660        0.491770  1.000001e+00       0.431205   
min         0.000000        0.000000 -1.841899e+00       0.000000   
25%         2.000000        0.000000 -6.850343e-01       1.000000   
50%         7.000000        0.000000 -2.422763e-01       1.000000   
75%        14.000000        1.000000  5.554445e-01       1.000000   
max        20.000000        1.000000  2.925911e+00       1.000000   

           Age_18-25      Age_26-35      Age_36-45      Age_46-50  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.181178       0.399200       0.199999       0.083082   
std         0.385166       0.489734       0.400000       0.276007   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

           Age_51-55        Age_55+  City_Category_B  City_Category_C  \
count  550068.000000  550068.000000    550068.000000    550068.000000   
mean        0.069993       0.039093         0.420263         0.311189   
std         0.255136       0.193817         0.493601         0.462980   
min         0.000000       0.000000         0.000000         0.000000   
25%         0.000000       0.000000         0.000000         0.000000   
50%         0.000000       0.000000         0.000000         0.000000   
75%         0.000000       0.000000         1.000000         1.000000   
max         1.000000       1.000000         1.000000         1.000000   

       Stay_In_Current_City_Years_1  Stay_In_Current_City_Years_2  \
count                 550068.000000                 550068.000000   
mean                       0.352358                      0.185137   
std                        0.477705                      0.388409   
min                        0.000000                      0.000000   
25%                        0.000000                      0.000000   
50%                        0.000000                      0.000000   
75%                        1.000000                      0.000000   
max                        1.000000                      1.000000   

       Stay_In_Current_City_Years_3  Stay_In_Current_City_Years_4+  \
count                 550068.000000                  550068.000000   
mean                       0.173224                       0.154028   
std                        0.378441                       0.360976   
min                        0.000000                       0.000000   
25%                        0.000000                       0.000000   
50%                        0.000000                       0.000000   
75%                        0.000000                       0.000000   
max                        1.000000                       1.000000   

          category_1     category_2     category_3     category_4  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.255201       0.132858       0.043104       0.071455   
std         0.435975       0.339422       0.203091       0.257583   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

          category_5     category_6     category_7     category_8  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.352367       0.076031       0.007903       0.346457   
std         0.477708      

In [ ]:
df_1.shape

(550068, 37)